crankers/src/vault_handler.rs

```diff
@@ -0,0 +1,410 @@
+                    filters: Some(vec![RpcFilterType::Memcmp(Memcmp::new(
+                        0,
+                        MemcmpEncodedBytes::Bytes(vec![VaultOperatorDelegation::DISCRIMINATOR]),
+                    ))]),
+                    account_config: RpcAccountInfoConfig {
+                        encoding: Some(UiAccountEncoding::Base64),
+                        ..RpcAccountInfoConfig::default()
+                    },
+                    ..RpcProgramAccountsConfig::default()
+                },
+            )
+            .await?;
+
+        let delegations: Vec<(Pubkey, VaultOperatorDelegation)> = accounts
+            .into_iter()
+            .filter_map(|(pubkey, acc)| {
+                VaultOperatorDelegation::try_from_slice_unchecked(&acc.data)
+                    .map_or(None, |v| Some((pubkey, *v)))
+            })
+            .collect();
+
+        Ok(delegations)
+    }
+
+    /// Retrieves the `VaultUpdateStateTracker` for a specific vault and epoch.
+    ///
+    /// # Returns
+    ///
+    /// Returns an `anyhow::Result<VaultUpdateStateTracker>` containing the deserialized state tracker
+    /// for the given vault and epoch. If successful, the state tracker is returned; otherwise,
+    /// an error is returned with contextual information.
+    pub async fn get_update_state_tracker(
+        &self,
+        vault: &Pubkey,
+        ncn_epoch: u64,
+    ) -> anyhow::Result<VaultUpdateStateTracker> {
+        let rpc_client = self.get_rpc_client();
+
+        let pubkey =
+            VaultUpdateStateTracker::find_program_address(&self.vault_program_id, vault, ncn_epoch)
+                .0;
+
+        match rpc_client.get_account(&pubkey).await {
+            Ok(account) => match VaultUpdateStateTracker::try_from_slice_unchecked(&account.data) {
+                Ok(tracker) => Ok(*tracker),
+                Err(e) => {
+                    let context = format!("Failed deserializing VaultUpdateStateTracker: {pubkey}");
+                    Err(anyhow::Error::new(e).context(context))
+                }
+            },
+            Err(e) => {
+                let context =
+                    format!("Error: Failed to get VaultUpdateStateTracker account: {pubkey}");
+                Err(anyhow::Error::new(e).context(context))
+            }
+        }
+    }
+
+    /// Performs a complete vault update cycle: initializes tracker, cranks it, and closes it.
+    ///
+    /// # Returns
+    ///
+    /// Returns `anyhow::Result<()>` indicating success or failure of the update operation.
+    pub async fn do_vault_update(
+        &self,
+        epoch: u64,
+        vault: &Pubkey,
+        operators: &[Pubkey],
+    ) -> anyhow::Result<()> {
+        let tracker_pubkey =
+            VaultUpdateStateTracker::find_program_address(&self.vault_program_id, vault, epoch).0;
+
+        log::info!("Updating vault: {vault}");
+
+        // Initialize
+        if self.get_update_state_tracker(vault, epoch).await.is_err() {
```

Comment:
can you log error here? it seems like there could be error if rpc down or the account doesn't exist (returns ForUser)